libcst/codemod/visitors/_apply_type_annotations.py

```diff
@@ -0,0 +1,483 @@
+            ),
+            *statements[1:],
+        ]
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> None:
+        self.qualifier.append(node.name.value)
+        self.visited_classes.add(node.name.value)
+
+    def leave_ClassDef(
+        self, original_node: cst.ClassDef, updated_node: cst.ClassDef
+    ) -> cst.ClassDef:
+        self.qualifier.pop()
+        return updated_node
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
+        self.qualifier.append(node.name.value)
+        # pyi files don't support inner functions, return False to stop the traversal.
+        return False
+
+    def leave_FunctionDef(
+        self, original_node: cst.FunctionDef, updated_node: cst.FunctionDef
+    ) -> cst.FunctionDef:
+        key = self._qualifier_name()
+        self.qualifier.pop()
+        if key in self.annotations.function_annotations:
+            function_annotation = self.annotations.function_annotations[key]
+            # Only add new annotation if one doesn't already exist
+            if not updated_node.returns:
+                updated_node = updated_node.with_changes(
+                    returns=function_annotation.returns
+                )
+            # Don't override default values when annotating functions
+            new_parameters = self._update_parameters(function_annotation, updated_node)
+            return updated_node.with_changes(params=new_parameters)
+        return updated_node
+
+    def leave_Assign(
+        self, original_node: cst.Assign, updated_node: cst.Assign
+    ) -> Union[cst.Assign, cst.AnnAssign]:
+
+        if len(original_node.targets) > 1:
+            for assign in original_node.targets:
+                target = assign.target
+                if isinstance(target, (cst.Name, cst.Attribute)):
+                    name = get_full_name_for_node(target)
+                    if name is not None:
+                        # Add separate top-level annotations for `a = b = 1`
+                        # as `a: int` and `b: int`.
+                        self._add_to_toplevel_annotations(name)
+            return updated_node
+        else:
+            return self._annotate_single_target(original_node, updated_node)
+
+    def leave_ImportFrom(
+        self, original_node: cst.ImportFrom, updated_node: cst.ImportFrom
+    ) -> cst.ImportFrom:
+        self.import_statements.append(original_node)
+        return updated_node
+
+    def visit_Comment(self, node: cst.Comment) -> None:
+        if "@" "generated" in node.value:
+            self.is_generated = True
+
+    def leave_Module(
+        self, original_node: cst.Module, updated_node: cst.Module
+    ) -> cst.Module:
+        fresh_class_definitions = [
+            definition
+            for name, definition in self.annotations.class_definitions.items()
+            if name not in self.visited_classes
+        ]
+        if self.is_generated:
+            return original_node
```

Comment:
`@generated` is FB specific logic which is not generic.
Our CLI includes the option to skip generated file.

https://github.com/Instagram/LibCST/blob/c21ae59a6bb2765d750748418c6bafc4af6cf8da/libcst/tool.py#L481

https://github.com/Instagram/LibCST/blob/c21ae59a6bb2765d750748418c6bafc4af6cf8da/libcst/tool.py#L657-L659

We should probably reuse those existing logic or at least make the current implementation configurable (e.g. add `skip_generated: bool = True` parameter to `add_stub_to_context`).

Reply:
Ok. I removed our ad hoc handling since it looks like your tool already supports it.